src/praisonai/praisonai/tool_resolver.py

```
@@ -342,32 +409,38 @@ def resolve(self, name: str, instantiate: bool = False) -> Optional[Callable]:
-            logger.debug(f"Resolved '{name}' from local tools.py")
-            tool = local_tools[name]
-        else:
-            # 2. Check wrapper ToolRegistry (ahead of SDK paths)
-            tool = self._resolve_from_wrapper_registry(name)
+        # Walk the configured source chain. Each source may return a plain
+        # callable / None, or a _ResolveResult to control cacheability. The
+        # first source that yields a non-None tool wins.
+        for source in self._sources:
+            try:
+                result = source.lookup(name)
+            except Exception as e:  # a misbehaving custom source must not break resolution
```

Ruff (0.15.18)

[warning] Do not catch blind exception: `Exception`

(BLE001)